### ISSUE TITLE
Explore integration of Anserini w/ Solr or Elasticsearch

### ISSUE THREAD

**Issue by lintool** [castorini/Anserini#364]
Solr or Elasticsearch are standard industry solutions for search, but their APIs are higher level than what's needed for research. This is where Anserini comes in. However, can we have our cake and eat it too?

By this, I mean: can we take an Anserini Lucene index and stick it into Solr or Elasticsearch directly, and have it be usable? That way, we can take advantage of standard user interfaces built on top of either search engine.

A quick search suggests that this is fairly straightforward with Solr: https://datafireball.com/2016/03/10/solr-load-lucene-index/

... but very difficult to do with Elasticsearch...

@r-clancy can you look into this and play around?

**Comment by r-clancy** [comment]
Came across this when doing some searching online: https://stackoverflow.com/questions/2715973/can-a-raw-lucene-index-be-loaded-by-solr

At least for Solr, it looks like it may be as simple as changing some fields in an XML file.

When I'm back in New Brunswick in a couple of weeks and have access to my PC, I'll play around with it a bit.

**Comment by tokee** [comment]
While will probably work fine to inject a Lucene index into Solr by stating the fields in the Solr `schema.xml`, there are a few gotchas.

1. You might have used an analyzer for Lucene in a way that cannot be configured in Solr. This is probably not the case, but it can happen.
2. If you have multiple Lucene indexes, you need to either merge them (probably not wise as we are talking web archive scale) or choose how to treat them in Solr. You might get away with specifying some custom routing and having a single collection, but the easiest solution is probably to create a separate collection for each index and use Solr's alias mechanism for doing aggregated search. Except for a bit of trickiness for distributed grouping, this will work fine - (it's what we do at the Royal Danish Library.

**Comment by lintool** [comment]
@tokee Thanks for your input!

Re: issue 2 - Can we build separate index shards and merge each one individually?

**Comment by tokee** [comment]
Yes, that is super easy. Lucene comes with an `IndexMergeTool` that does just what it says on the tin.

There is a hard limit of 2 billion documents/index and the same limit is found internally for a lot of structures. We cap our sub-collections (indexes really) at 300M records. We might get away 500M, but we cannot go much higher as the number of outgoing links (which we index in a designated field) would exceed 2 billion.

**Comment by tokee** [comment]
If your question was if you can merge the indexes into Solr one at a time, then the answer is still yes if you treat them as separate collections: You create a new (empty) collection in Solr with the Lucene-compatible schema, copy the Lucene index files into the data folder and reload the Solr core. Then you add the collection to the alias for all active collections.

**Comment by lintool** [comment]
okay, thanks for the helpful tips!

**Comment by r-clancy** [comment]
I've had some time to look into this, it's pretty straightforward to use a Lucene index in Solr. 

I've document the process here - https://www.notion.so/rclancy/Solr-8931d6fdabd447a2996a0a42d34e5895

**Comment by lintool** [comment]
Sync'ed up with Ryan, clarified a few points.

Try http://projectblacklight.org/ as a nice front end.

**Comment by tokee** [comment]
@r-clancy I presume you have more than 2 fields (`id` and `content`) in your index? When using Blacklight or (shameless company plug) SolrWayback, you might want/need to experiment with [field name aliasing](https://lucene.apache.org/solr/guide/7_0/common-query-parameters.html#field-name-aliases) to match your fields match as best as possible to the fields defined in the [webarchive-discovery schema](https://github.com/ukwa/webarchive-discovery/blob/master/warc-indexer/src/main/solr/solr7/discovery/conf/schema.xml).

**Comment by r-clancy** [comment]
@tokee The data I've been testing with only has two fields (`id` and `contents`), but I'll make a note of that for the future - thanks!

**Comment by r-clancy** [comment]
I've had some time to look into this a bit more.

I've updated the [doc](https://www.notion.so/rclancy/Solr-8931d6fdabd447a2996a0a42d34e5895) to use `type="text_en"` which more closely resembles the stemming used in Anserini.

I've created a docker image for Project Blacklight (which I'll add to the document in the next day or two) that starts up an instance of Blacklight so we can import data into the embedded Solr instance and run queries on the front-end. It's has lots of customization options (i.e. the defaults don't work with our data) so once the config has been updated to point to the proper fields it should work great (basic queries can be done with the current setup).

Next steps are to validate the queries return the same results (in the command line experiments vs. Solr queries) and clean up the Blacklight docker image and configuration.
